dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/authorize/AbstractScopeAuthorizationStrategy.java

```diff
@@ -0,0 +1,113 @@
+ * this program. If not, see <http://www.gnu.org/licenses/>.                                                     
+ *                                                                                                               
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY                           
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES                          
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT                           
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,                                
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED                          
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;                               
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER                              
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN                         
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.icgc.dcc.storage.server.authorize;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import lombok.NonNull;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+import org.icgc.dcc.storage.server.service.MetadataService;
+import org.icgc.dcc.storage.server.util.AuthScope;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+
+@Slf4j
+public abstract class AbstractScopeAuthorizationStrategy {
+
+  protected AuthScope scope;
+
+  @Autowired
+  protected MetadataService metadataClient;
+
+  public AbstractScopeAuthorizationStrategy() {
+    super();
+  }
+
+  /**
+   * Package-private constructor for use in unit tests (to initialize scope context) without a Spring Context
+   */
+  protected AbstractScopeAuthorizationStrategy(final String scopeStr) {
+    setAuthorizeScope(scopeStr);
+    scope = AuthScope.from(getAuthorizeScope());
+  }
+
+  protected abstract void setAuthorizeScope(String scopeStr);
+
+  protected abstract String getAuthorizeScope();
+
+  protected abstract boolean verify(@NonNull List<AuthScope> grantedScopes, @NonNull final String objectId);
+
+  /**
+   * Main entry point: version of authorize() method that assumes object id is passed-in (method-level security has
+   * access to this)
+   */
+  public boolean authorize(@NonNull Authentication authentication, @NonNull final String objectId) {
+    log.info("Checking authorization with object id {}", objectId);
+    scope = AuthScope.from(getAuthorizeScope());
+
+    // TODO: urrrrr
```

Comment:
I originally tried having something like a getGrantedScopes(OAuth2Authentication auth) and getGrantedScopes(OtherAuthenticationType auth) but it wasn't resolving correctly.